text/0126-Configurable-Metric-Aggregations.md

```diff
@@ -0,0 +1,97 @@
+# A Proposal For SDK Support for Configurable Batching and Aggregations
+
+Add support to the default SDK for the ability to configure Metric Aggregations.
+
+## Motivation
+
+OpenTelemetry's architecture separates the concerns of instrumentation and operation. The Metric Instruments
+provided by the Metric API are all defined to have a default aggregation. And, by default, aggregations are
+performed with all Labels being used to define a unit of aggregation. Although this is a good default
+configuration for the SDK to provide, more configurability is needed.
+
+There are 3 main use-cases that this proposal is intended to address:
+
+1) The application developer/operator wishes to use an aggregation other than the default provided by the SDK
+for a given instrument or set of instruments.
+2) An exporter author wishes to inform the SDK what "Temporality" (delta vs. cumulative) the resulting metric
```

Comment:
Don't see how this PR is explaining how this is achieved. There is now way (in this PR) for an ExporterAuthor to tell the SDK that it expects delta or cumulative, right? The end user must know whether the exporter he is using is expecting delta or cumulative, and use that knowledge to setup Temporality right?

Reply:
This is a good point. I had been assuming that exporters could get access to the "global" SDK instance the same way that any other SDK user could. But, if there is no global SDK in a given language implementation, we may need to provide a mechanism for exporters to get access explicitly to the SDK that they are being used by. I think this is related to the open question about whether Temporality is something that should only be specifiable by an exporter.

Reply:
I added an open question to capture this.

Reply:
Ok. I'd prefer if Exporters don't have to retrieve a Global SDK, and change its settings. Exporter can expose a method like `GetExpectedTemporality()`, and end user can read this property from the exporter, and use that to set the SDK config. So that exporters do what it is supposed to do - export, and not do anything with SDK settings/config.

Anyway - this is not blocking the PR. I approve.

Reply:
FYI I solved this problem in https://github.com/open-telemetry/opentelemetry-go/pull/840 for OTel-Go as follows. 
1. The Processor gets an ExportKindSelector for determining the ExportKind of a specific descriptor
2. The Exporter implements ExportKindSelector with its own preferences
3. CheckpointSet.ForEach requires you to pass the ExportKindSelector again

The processor calls the ExportKindSelector when it encounters new metrics. The exporter calls ForEach with the specific export it wants to view. The selector passed in (1) can return multiple kinds to allow multiple sub-exporters. The selector passed in (3) must return a single kind when reading values in the exporter.

```
// Exporter handles presentation of the checkpoint of aggregate
// metrics.  This is the final stage of a metrics export pipeline,
// where metric data are formatted for a specific system.
type Exporter interface {
	// Export is called immediately after completing a collection
	// pass in the SDK.
	//
	// The Context comes from the controller that initiated
	// collection.
	//
	// The CheckpointSet interface refers to the Processor that just
	// completed collection.
	Export(context.Context, CheckpointSet) error

	// ExportKindSelector is an interface used by the Processor
	// in deciding whether to compute Delta or Cumulative
	// Aggregations when passing Records to this Exporter.
	ExportKindSelector
}

// ExportKindSelector is a sub-interface of Exporter used to indicate
// whether the Processor should compute Delta or Cumulative
// Aggregations.
type ExportKindSelector interface {
	// ExportKindFor should return the correct ExportKind that
	// should be used when exporting data for the given metric
	// instrument and Aggregator kind.
	ExportKindFor(*metric.Descriptor, aggregation.Kind) ExportKind
}

// CheckpointSet allows a controller to access a complete checkpoint of
// aggregated metrics from the Processor.  This is passed to the
// Exporter which may then use ForEach to iterate over the collection
// of aggregated metrics.
type CheckpointSet interface {
	// ForEach iterates over aggregated checkpoints for all
	// metrics that were updated during the last collection
	// period. Each aggregated checkpoint returned by the
	// function parameter may return an error.
	//
	// The ExportKindSelector argument is used to determine
	// whether the Record is computed using Delta or Cumulative
	// aggregation.
	//
	// ForEach tolerates ErrNoData silently, as this is
	// expected from the Meter implementation. Any other kind
	// of error will immediately halt ForEach and return
	// the error to the caller.
	ForEach(ExportKindSelector, func(Record) error) error

	// Locker supports locking the checkpoint set.  Collection
	// into the checkpoint set cannot take place (in case of a
	// stateful processor) while it is locked.
	//
	// The Processor attached to the Accumulator MUST be called
	// with the lock held.
	sync.Locker

	// RLock acquires a read lock corresponding to this Locker.
	RLock()
	// RUnlock releases a read lock corresponding to this Locker.
	RUnlock()
}
```